cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamLeaderManager.java

```diff
@@ -0,0 +1,217 @@
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final StreamMetaStore streamMetaStore;
+  private final ResourceCoordinatorClient streamsResourceCoordinatorClient;
+
+  private LeaderElection leaderElection;
+  private ResourceCoordinator streamsResourceCoordinator;
+  private Cancellable handlerSubscription;
+
+  @Inject
+  public DistributedStreamLeaderManager(CConfiguration cConf, ZKClient zkClient,
+                                        DiscoveryServiceClient discoveryServiceClient,
+                                        StreamMetaStore streamMetaStore) {
+    this.cConf = cConf;
+    this.zkClient = zkClient;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.streamMetaStore = streamMetaStore;
+    this.streamsResourceCoordinatorClient = new ResourceCoordinatorClient(zkClient);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    streamsResourceCoordinatorClient.startAndWait();
+
+    // Start the resource coordinator that will map Streams to Stream handlers
+    leaderElection = new LeaderElection(
+      zkClient, "/election/" + Constants.Service.STREAMS_COORDINATOR, new ElectionHandler() {
+      @Override
+      public void leader() {
+        LOG.info("Became Stream handler leader. Starting resource coordinator.");
+        streamsResourceCoordinator = new ResourceCoordinator(zkClient, discoveryServiceClient,
+                                                             new BalancedAssignmentStrategy());
+        streamsResourceCoordinator.startAndWait();
+
+        try {
+          // Create one requirement for the resource coordinator for all the streams.
+          // One stream is identified by one partition
+          ResourceRequirement.Builder builder = ResourceRequirement.builder(STREAMS_RESOURCE);
+          for (StreamSpecification spec : streamMetaStore.listStreams()) {
+            LOG.debug("Adding {} stream as a resource to the coordinator to manager streams leaders.", spec.getName());
+            builder.addPartition(new ResourceRequirement.Partition(spec.getName(), 1));
+          }
+          streamsResourceCoordinatorClient.submitRequirement(builder.build()).get();
+        } catch (Throwable e) {
+          LOG.error("Could not create requirement for coordinator in Stream handler leader", e);
+          Throwables.propagate(e);
+        }
+      }
+
+      @Override
+      public void follower() {
+        LOG.info("Became Stream handler follower.");
+        if (streamsResourceCoordinator != null) {
+          streamsResourceCoordinator.stopAndWait();
+        }
+      }
+    });
+
+    // All handlers subscribe to the resource coordinator
+    // Discoverables here have to be unique, don't care about the name and port but for comparison
+    final int instanceId = cConf.getInt(Constants.Stream.CONTAINER_INSTANCE_ID, 1);
+    Discoverable discoverable = createDiscoverable(String.format("stream-handler-%d", instanceId), 30000 + instanceId);
```

Comment:
Although we don't need a real port, but it's a better idea to use the Stream http service port.